src/ray/raylet/node_manager.cc

```diff
@@ -130,6 +137,30 @@ ray::Status NodeManager::RegisterGcs() {
       JobID::nil(), gcs_client_->client_table().GetLocalClientId(),
       task_committed_callback, nullptr, nullptr));
 
+  const auto task_lease_notification_callback = [this](gcs::AsyncGcsClient *client,
+                                                       const TaskID &task_id,
+                                                       const TaskLeaseDataT &task_lease) {
```

Comment:
E.g., I wonder if it should all be done in the `ReconstructionPolicy`, e.g., the first bullet could be done in the constructor and the third could be addressed since the `TaskDependencyManager` has a handle to the `ReconstructionPolicy` anyway.

Reply:
Yeah, I agree it's not great that the task leases are spread throughout multiple components. On the other hand, I sort of like the current separation of logic between the `TaskDependencyManager` and `ReconstructionPolicy`.

I do agree that bullet 1 should be done in the `ReconstructionPolicy`. I might leave this for a future PR, though, since it may require some changes to the `AsyncGcsClient` abstraction.